File: tests/oracle/fspec_spaceless_rem_1204.cc
/*
 * FSPEC-SPACELESS-REM-1204: locked Ghidra 12.0.4 oracle for the
 * FSPEC-SPACELESS-REMAINDER fix — the last two spaceless
 * justifiedContain callers:
 *
 *  - case=uc_plain_cross_space / uc_join_reachable:
 *    ParamListStandard::unjustifiedContainer (fspec.cc:1411-1424)
 *    iterates ALL entries with NO caller-level space filter; space
 *    rejection happens only INSIDE justifiedContain (fspec.cc:269 /
 *    address.cc:133 for plain entries; per-piece address.cc:133 for
 *    joins). A stack query at a numerically unjustified register/ram
 *    offset returns hit=0 (the old spaceless arithmetic would have
 *    returned the register container); join entries ARE reachable and
 *    getContainer (fspec.cc:295) passes back the containing PIECE
 *    (register:0x100/4), not the whole join.
 *  - case=fb_plain_best_cover / fb_join_reachable / fb_cross_space_join /
 *    fb_first_only_skip / fb_first_only_allowed:
 *    ParamListStandardOut::fillinMapFallback (fspec.cc:1638-1719): both
 *    trial queries (cc:1656 evaluate-in-terms-of-current-entry and
 *    cc:1702 best-entry re-evaluation) are space-aware — the trial
 *    Address carries its space, join entries match register/ram trials
 *    through the per-piece walk (the old caller-level getSpace()
 *    equality guard made every join entry unreachable -> bestentry
 *    null -> all trials markNoUse), the best loop overwrites the
 *    per-entry evaluation state, the offmatch contiguity walk, the
 *    minSize coverage gate, the type/cover preference
 *    (TYPECLASS_GENERAL < TYPECLASS_PTR), and the firstOnly skip
 *    (cc:1649-1652: !isFirstInClass && isExclusion &&
 *    getAllGroups().size() == 1).
 *
 * ParamEntry/ParamListStandard(out)/ParamActive/ParamTrial construction
 * uses the same class->struct access hack as the
 * fspec_possibleparam_1204 fixture; every observation goes through
 * public members only. fillinMapFallback never touches the resolver, so
 * populateResolver is intentionally NOT called.
 */

#include <bits/stdc++.h>

#define class struct
#define private public
#define protected public
#include "fspec.hh"
#include "opcodes.hh"
#include "space.hh"
#include "translate.hh"
#undef protected
#undef private
#undef class

using namespace ghidra;

class FixtureTranslate final : public Translate {
public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
  }

  std::string tryInsertSpace(AddrSpace *spc) {
    try {
      insertSpace(spc);
      return "ok";
    }
    catch (LowlevelError &err) {
      return std::string("err ") + err.explain;
    }
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    static VarnodeData dummy;
    return dummy;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

// canonical synthetic spaces: constant=0, other=1, unique=2, ram=3,
// register=4, stack=5, join=6 (same index plan as the
// fspec_findentry_1204 / fspec_possibleparam_1204 fixtures).
static void buildSpaces(FixtureTranslate *tr, AddrSpace *&constant,
                        AddrSpace *&unique, AddrSpace *&ram,
                        AddrSpace *&reg, AddrSpace *&stack, AddrSpace *&join) {
  tr->tryInsertSpace(new ConstantSpace(tr, tr));
  tr->tryInsertSpace(new OtherSpace(tr, tr, OtherSpace::INDEX));
  tr->tryInsertSpace(new UniqueSpace(tr, tr, 2, 0));
  ram = new AddrSpace(tr, tr, IPTR_PROCESSOR, "ram", false, 8, 1, 3,
                      AddrSpace::hasphysical, 0, 0);
  reg = new AddrSpace(tr, tr, IPTR_PROCESSOR, "register", false, 8, 1, 4,
                      AddrSpace::hasphysical, 0, 0);
  stack = new AddrSpace(tr, tr, IPTR_SPACEBASE, "stack", false, 8, 1, 5,
                        AddrSpace::hasphysical, 0, 0);
  tr->tryInsertSpace(ram);
  tr->tryInsertSpace(reg);
  tr->tryInsertSpace(stack);
  constant = tr->getConstantSpace();
  unique = tr->getUniqueSpace();
  join = new JoinSpace(tr, tr, 6);
}

// Fill ParamEntry fields directly (staged cspec loader, same pattern
// as the fspec_findentry_1204 fixture).
static ParamEntry makeEntry(int4 grp, AddrSpace *spc, uintb base, int4 size,
                            int4 minsize, int4 alignment, uint4 flags) {
  ParamEntry e(grp);
  e.flags = flags;
  e.type = TYPECLASS_GENERAL;
  e.groupSet.clear();
  e.groupSet.push_back(grp);
  e.spaceid = spc;
  e.addressbase = base;
  e.size = size;
  e.minsize = minsize;
  e.alignment = alignment;
  e.numslots = (alignment != 0) ? (size / alignment) : 1;
  e.joinrec = (JoinRecord *)0;
  return e;
}

// Stage a join ParamEntry (spaceid = join space, pieces MOST
// significant first), the state resolveJoin caches after decoding a
// join <pentry>.
static ParamEntry makeJoinEntry(int4 grp, AddrSpace *joinspc, JoinRecord *rec,
                                int4 size, int4 minsize) {
  ParamEntry e(grp);
  e.flags = 0;
  e.type = TYPECLASS_GENERAL;
  e.groupSet.clear();
  e.groupSet.push_back(grp);
  e.spaceid = joinspc;
  e.addressbase = 0;
  e.size = size;
  e.minsize = minsize;
  e.alignment = 0;
  e.numslots = 1;
  e.joinrec = rec;
  return e;
}

static void printUc(ParamListStandard *stds, const char *spcname,
                    AddrSpace *spc, uintb off, int4 sz) {
  VarnodeData res;
  res.space = (AddrSpace *)0;
  res.offset = 0;
  res.size = 0;
  bool hit = stds->unjustifiedContainer(Address(spc, off), sz, res);
  std::cout << "  uc spc=" << spcname << " off=0x" << std::hex << off
            << " sz=" << std::dec << sz << " hit=" << (hit ? 1 : 0);
  if (hit)
    std::cout << " res=" << res.space->getName() << ":0x" << std::hex
              << res.offset << "/" << std::dec << res.size;
  std::cout << std::endl;
}

// Map the trial's ParamEntry pointer back to its list index (-1 null).
static int4 entryIndex(const std::vector<const ParamEntry *> &ptrs,
                       const ParamTrial &t) {
  const ParamEntry *ent = t.getEntry();
  if (ent == (const ParamEntry *)0) return -1;
  for (int4 i = 0; i < (int4)ptrs.size(); ++i)
    if (ptrs[i] == ent) return i;
  return -2; // unreachable in this fixture
}

// Register active trials then run fillinMapFallback and dump the
// final trial order (post sortTrials) with used/entry/offset state.
static void runFb(ParamListStandardOut *out, ParamActive *active,
                  const char * /*label*/, bool firstOnly) {
  out->fillinMapFallback(active, firstOnly);
  std::vector<const ParamEntry *> ptrs;
  for (list<ParamEntry>::const_iterator it = out->entry.begin();
       it != out->entry.end(); ++it)
    ptrs.push_back(&(*it));
  for (int4 i = 0; i < active->getNumTrials(); ++i) {
    const ParamTrial &t(active->getTrial(i));
    std::cout << "  fb spc=" << t.getAddress().getSpace()->getName()
              << " off=0x" << std::hex << t.getAddress().getOffset()
              << " sz=" << std::dec << t.getSize()
              << " used=" << (t.isUsed() ? 1 : 0)
              << " entry=" << entryIndex(ptrs, t)
              << " joff=" << t.getOffset() << std::endl;
  }
}

int main(void) {
  std::cout << "schema=1|fixture=FSPEC-SPACELESS-REM-1204|"
               "oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
            << std::endl;

  FixtureTranslate tr;
  AddrSpace *constant, *unique, *ram, *reg, *stack, *join;
  buildSpaces(&tr, constant, unique, ram, reg, stack, join);

  // ---- case 1: plain entries, cross-space rejection ----------------
  std::cout << "case=uc_plain_cross_space" << std::endl;
  {
    ParamListStandard stds;
    // e0: register [0x100,0x107] min 2, exclusion, force-left
    stds.entry.push_back(makeEntry(0, reg, 0x100, 8, 2, 0,
                                   ParamEntry::force_left_justify));
    // e1: ram [0x100,0x10F] min 4, alignment 8 (non-exclusion)
    stds.entry.push_back(makeEntry(1, ram, 0x100, 16, 4, 8, 0));
    // e2: register [0x200,0x207] min 4, exclusion, force-left
    stds.entry.push_back(makeEntry(2, reg, 0x200, 8, 4, 0,
                                   ParamEntry::force_left_justify));
    stds.numgroup = 3;
    struct UcCase { const char *spcname; AddrSpace *spc; uintb off; int4 sz; };
    const UcCase uc[] = {
      { "reg", reg, 0x104, 4 },   // unjustified in e0 -> container 0x100/8
      { "reg", reg, 0x100, 4 },   // just==0 -> early false
      { "stack", stack, 0x104, 4 }, // address.cc:133 -> hit=0
      { "ram", ram, 0x104, 8 },   // e1 aligned route, just=4 -> 0x100/16
      { "reg", reg, 0x100, 1 },   // minSize gate (2 > 1, 4 > 1)
      { "ram", ram, 0x108, 8 },   // aligned justified -> just==0 -> false
      { "reg", reg, 0x202, 2 },   // e2 container 0x200/8
    };
    for (const UcCase &c : uc)
      printUc(&stds, c.spcname, c.spc, c.off, c.sz);
  }

  // ---- case 2: join entries ARE reachable --------------------------
  std::cout << "case=uc_join_reachable" << std::endl;
  {
    JoinRecord rec;
    VarnodeData high, low;
    high.space = ram; high.offset = 0x204; high.size = 4;
    low.space = reg; low.offset = 0x100; low.size = 4;
    rec.pieces.push_back(high);
    rec.pieces.push_back(low);
    ParamListStandard stds;
    // e0: join ram:0x204 (high) + reg:0x100 (low)
    stds.entry.push_back(makeJoinEntry(0, join, &rec, 8, 4));
    // e1: register [0x100,0x107] min 4, exclusion, force-left
    stds.entry.push_back(makeEntry(1, reg, 0x100, 8, 4, 0,
                                   ParamEntry::force_left_justify));
    stds.numgroup = 2;
    struct UcCase { const char *spcname; AddrSpace *spc; uintb off; int4 sz; };
    const UcCase uc[] = {
      { "reg", reg, 0x102, 2 },   // low piece just=2 -> PIECE container reg:0x100/4
      { "reg", reg, 0x104, 4 },   // join walk -1 (poke out + foreign high) -> e1 just=4
      { "reg", reg, 0x204, 4 },   // both pieces foreign -> -1 (numeric high hit rejected)
      { "ram", ram, 0x204, 4 },   // skip low (+4), high cur=0 -> 4 -> PIECE ram:0x204/4
      { "reg", reg, 0x100, 4 },   // low piece just=0 -> early false
      { "stack", stack, 0x102, 2 }, // per-piece address.cc:133 -> hit=0
    };
    for (const UcCase &c : uc)
      printUc(&stds, c.spcname, c.spc, c.off, c.sz);
  }

  // ---- case 3: fallback, plain entries, best cover ------------------
  std::cout << "case=fb_plain_best_cover" << std::endl;
  {
    ParamListStandardOut out;
    out.entry.push_back(makeEntry(0, reg, 0x100, 8, 4, 0,
                                  ParamEntry::force_left_justify |
                                  ParamEntry::first_storage));
    out.entry.push_back(makeEntry(1, ram, 0x100, 16, 4, 8,
                                  ParamEntry::first_storage));
    // e2: same range as e0 but min 8 — coverage tie rejected
    out.entry.push_back(makeEntry(2, reg, 0x100, 8, 8, 0,
                                  ParamEntry::force_left_justify |
                                  ParamEntry::first_storage));
    out.numgroup = 3;
    ParamActive active(false);
    active.registerTrial(Address(reg, 0x104), 4);
    active.registerTrial(Address(reg, 0x100), 4);
    active.registerTrial(Address(stack, 0x104), 4);
    for (int4 i = 0; i < active.getNumTrials(); ++i)
      active.getTrial(i).markActive();
    runFb(&out, &active, "plain", false);
  }

  // ---- case 4: fallback, join entry reachable (core divergence) -----
  std::cout << "case=fb_join_reachable" << std::endl;
  {
    JoinRecord rec;
    VarnodeData high, low;
    high.space = reg; high.offset = 0x104; high.size = 4;
    low.space = reg; low.offset = 0x100; low.size = 4;
    rec.pieces.push_back(high);
    rec.pieces.push_back(low);
    ParamListStandardOut out;
    out.entry.push_back(makeJoinEntry(0, join, &rec, 8, 4));
    out.entry.back().flags |= ParamEntry::first_storage;
    out.entry.push_back(makeEntry(1, reg, 0x200, 8, 4, 0,
                                  ParamEntry::force_left_justify |
                                  ParamEntry::first_storage));
    out.numgroup = 2;
    ParamActive active(false);
    active.registerTrial(Address(reg, 0x100), 4);
    active.registerTrial(Address(reg, 0x104), 4);
    for (int4 i = 0; i < active.getNumTrials(); ++i)
      active.getTrial(i).markActive();
    runFb(&out, &active, "join", false);
  }

  // ---- case 5: fallback, cross-space join pieces ---------------------
  std::cout << "case=fb_cross_space_join" << std::endl;
  {
    JoinRecord rec;
    VarnodeData high, low;
    high.space = ram; high.offset = 0x204; high.size = 4;
    low.space = reg; low.offset = 0x100; low.size = 4;
    rec.pieces.push_back(high);
    rec.pieces.push_back(low);
    ParamListStandardOut out;
    out.entry.push_back(makeJoinEntry(0, join, &rec, 8, 4));
    out.entry.back().flags |= ParamEntry::first_storage;
    out.numgroup = 1;
    ParamActive active(false);
    active.registerTrial(Address(reg, 0x100), 4);
    active.registerTrial(Address(ram, 0x204), 4);
    active.registerTrial(Address(stack, 0x204), 4);
    for (int4 i = 0; i < active.getNumTrials(); ++i)
      active.getTrial(i).markActive();
    runFb(&out, &active, "cross", false);
  }

  // ---- case 6: firstOnly skips non-first single-group exclusions -----
  std::cout << "case=fb_first_only_skip" << std::endl;
  {
    ParamListStandardOut out;
    out.entry.push_back(makeEntry(0, reg, 0x100, 8, 4, 0,
                                  ParamEntry::force_left_justify |
                                  ParamEntry::first_storage));
    // e1: NOT first_storage, exclusion (alignment 0), one group
    out.entry.push_back(makeEntry(1, reg, 0x108, 8, 4, 0,
                                  ParamEntry::force_left_justify));
    out.numgroup = 2;
    ParamActive active(false);
    active.registerTrial(Address(reg, 0x108), 4);
    active.getTrial(0).markActive();
    runFb(&out, &active, "skip", true);
  }

  // ---- case 7: same layout, firstOnly=false reaches e1 ----------------
  std::cout << "case=fb_first_only_allowed" << std::endl;
  {
    ParamListStandardOut out;
    out.entry.push_back(makeEntry(0, reg, 0x100, 8, 4, 0,
                                  ParamEntry::force_left_justify |
                                  ParamEntry::first_storage));
    out.entry.push_back(makeEntry(1, reg, 0x108, 8, 4, 0,
                                  ParamEntry::force_left_justify));
    out.numgroup = 2;
    ParamActive active(false);
    active.registerTrial(Address(reg, 0x108), 4);
    active.getTrial(0).markActive();
    runFb(&out, &active, "allowed", false);
  }

  return 0;
}

File: tests/oracle/fspec_spaceless_rem_1204.rs
// FSPEC-SPACELESS-REM-1204: Rugra comparand for the locked Ghidra
// 12.0.4 oracle (FSPEC-SPACELESS-REMAINDER). Mirrors
// tests/oracle/fspec_spaceless_rem_1204.cc case for case:
//  - uc_plain_cross_space / uc_join_reachable:
//    ParamListStandard::unjustified_container (fspec.cc:1411-1424)
//    threads the query space into justified_contain_in_space — no
//    caller-level filter, space rejection only inside the walk
//    (fspec.cc:269 / address.cc:133 plain; per-piece address.cc:133
//    joins, which ARE reachable; getContainer passes back the
//    containing PIECE, fspec.cc:295).
//  - fb_*: ParamListStandardOut::fillin_map_fallback (fspec.cc:1638-
//    1719) — both trial queries (cc:1656 and the cc:1702 best-entry
//    re-evaluation) go through justified_contain_in_space with the
//    trial's own space; the caller-level getSpace() equality guard is
//    gone, join entries match register/ram trials per piece, the best
//    loop overwrites per-entry evaluation state, the offmatch
//    contiguity walk, minSize coverage gate, type/cover preference,
//    and the firstOnly skip (cc:1649-1652) are exercised.

use rugra::address::Address;
use rugra::fspec::param_entry_flags;
use rugra::fspec::{ParamActive, ParamEntry, ParamListStandard,
                   ParamListStandardOut, TypeClass, VarnodeData};
use rugra::space::AddressSpace;

// staged-loader entry builder (the pub equivalent of the C++ fixture's
// direct field writes; same pattern as the fspec_possibleparam_1204
// fixture).
fn make_entry(
    grp: i32,
    spc: AddressSpace,
    base: u64,
    size: i32,
    min_size: i32,
    alignment: i32,
    flags: u32,
) -> ParamEntry {
    let mut e = ParamEntry::new(grp);
    *e.flags_mut() = flags;
    e.set_type_class(TypeClass::General);
    e.set_space(spc);
    e.set_base(base);
    e.set_sizes(size, min_size);
    e.set_alignment(alignment); // normalizes alignment == size to 0
    e
}

// Stage a join ParamEntry (space = join space + pieces, most
// significant first) — the state resolveJoin caches after decoding a
// join <pentry>.
fn make_join_entry(
    grp: i32,
    size: i32,
    min_size: i32,
    pieces: &[VarnodeData],
) -> ParamEntry {
    let mut e = ParamEntry::new(grp);
    e.set_type_class(TypeClass::General);
    e.set_space(AddressSpace::Join);
    e.set_base(0);
    e.set_sizes(size, min_size);
    e.set_alignment(0);
    e.set_join_pieces(pieces.to_vec());
    e
}

fn print_uc(model: &ParamListStandard, spcname: &str, space: AddressSpace,
            off: u64, sz: i32) {
    let mut res = VarnodeData { space: AddressSpace::Ram, offset: 0, size: 0 };
    let hit = model.unjustified_container(space, Address::new(off), sz, &mut res);
    let mut line = format!("  uc spc={} off=0x{:x} sz={} hit={}", spcname, off, sz,
                           if hit { 1 } else { 0 });
    if hit {
        line.push_str(&format!(" res={}:0x{:x}/{}", res.space.name(),
                               res.offset, res.size));
    }
    println!("{}", line);
}

// Register active trials then run fillin_map_fallback and dump the
// final trial order (post sort_trials) with used/entry/offset state.
fn run_fb(out: &ParamListStandardOut, active: &mut ParamActive, first_only: bool) {
    out.fillin_map_fallback(active, first_only);
    for i in 0..active.get_num_trials() {
        let t = active.get_trial(i);
        let entry = t.get_entry_index().map(|ix| ix as i32).unwrap_or(-1);
        println!("  fb spc={} off=0x{:x} sz={} used={} entry={} joff={}",
                 t.get_space().name(), t.get_address().as_u64(), t.get_size(),
                 t.is_used() as i32, entry, t.get_offset());
    }
}

fn main() {
    println!("schema=1|fixture=FSPEC-SPACELESS-REM-1204|oracle=e40ed13014025f82488b1f8f7bca566894ac376b");

    // ---- case 1: plain entries, cross-space rejection ----------------
    println!("case=uc_plain_cross_space");
    {
        let mut model = ParamListStandard::new();
        // e0: register [0x100,0x107] min 2, exclusion, force-left
        model.entry_mut().push(make_entry(
            0, AddressSpace::Register, 0x100, 8, 2, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        // e1: ram [0x100,0x10F] min 4, alignment 8 (non-exclusion)
        model.entry_mut().push(make_entry(
            1, AddressSpace::Ram, 0x100, 16, 4, 8, 0,
        ));
        // e2: register [0x200,0x207] min 4, exclusion, force-left
        model.entry_mut().push(make_entry(
            2, AddressSpace::Register, 0x200, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        model.set_num_group(3);
        let uc: [(&str, AddressSpace, u64, i32); 7] = [
            ("reg", AddressSpace::Register, 0x104, 4),   // unjustified in e0 -> 0x100/8
            ("reg", AddressSpace::Register, 0x100, 4),   // just==0 -> early false
            ("stack", AddressSpace::Stack, 0x104, 4),    // address.cc:133 -> hit=0
            ("ram", AddressSpace::Ram, 0x104, 8),        // e1 aligned route, just=4 -> 0x100/16
            ("reg", AddressSpace::Register, 0x100, 1),   // minSize gate
            ("ram", AddressSpace::Ram, 0x108, 8),        // aligned justified -> false
            ("reg", AddressSpace::Register, 0x202, 2),   // e2 container 0x200/8
        ];
        for &(name, space, off, sz) in uc.iter() {
            print_uc(&model, name, space, off, sz);
        }
    }

    // ---- case 2: join entries ARE reachable --------------------------
    println!("case=uc_join_reachable");
    {
        let pieces = [
            VarnodeData { space: AddressSpace::Ram, offset: 0x204, size: 4 },
            VarnodeData { space: AddressSpace::Register, offset: 0x100, size: 4 },
        ];
        let mut model = ParamListStandard::new();
        // e0: join ram:0x204 (high) + reg:0x100 (low)
        model.entry_mut().push(make_join_entry(0, 8, 4, &pieces));
        // e1: register [0x100,0x107] min 4, exclusion, force-left
        model.entry_mut().push(make_entry(
            1, AddressSpace::Register, 0x100, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        model.set_num_group(2);
        let uc: [(&str, AddressSpace, u64, i32); 6] = [
            ("reg", AddressSpace::Register, 0x102, 2),   // low piece just=2 -> PIECE reg:0x100/4
            ("reg", AddressSpace::Register, 0x104, 4),   // join walk -1 -> e1 just=4
            ("reg", AddressSpace::Register, 0x204, 4),   // both pieces foreign -> -1
            ("ram", AddressSpace::Ram, 0x204, 4),        // skip low (+4), high cur=0 -> PIECE ram:0x204/4
            ("reg", AddressSpace::Register, 0x100, 4),   // low piece just=0 -> early false
            ("stack", AddressSpace::Stack, 0x102, 2),    // per-piece address.cc:133 -> hit=0
        ];
        for &(name, space, off, sz) in uc.iter() {
            print_uc(&model, name, space, off, sz);
        }
    }

    // ---- case 3: fallback, plain entries, best cover ------------------
    println!("case=fb_plain_best_cover");
    {
        let mut out = ParamListStandardOut::new();
        out.base.entry_mut().push(make_entry(
            0, AddressSpace::Register, 0x100, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY
                | param_entry_flags::FIRST_STORAGE,
        ));
        out.base.entry_mut().push(make_entry(
            1, AddressSpace::Ram, 0x100, 16, 4, 8,
            param_entry_flags::FIRST_STORAGE,
        ));
        // e2: same range as e0 but min 8 — coverage tie rejected
        out.base.entry_mut().push(make_entry(
            2, AddressSpace::Register, 0x100, 8, 8, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY
                | param_entry_flags::FIRST_STORAGE,
        ));
        out.base.set_num_group(3);
        let mut active = ParamActive::new(false);
        active.register_trial_in_space(AddressSpace::Register, Address::new(0x104), 4);
        active.register_trial_in_space(AddressSpace::Register, Address::new(0x100), 4);
        active.register_trial_in_space(AddressSpace::Stack, Address::new(0x104), 4);
        for i in 0..active.get_num_trials() {
            active.get_trial_mut(i).mark_active();
        }
        run_fb(&out, &mut active, false);
    }

    // ---- case 4: fallback, join entry reachable (core divergence) -----
    println!("case=fb_join_reachable");
    {
        let pieces = [
            VarnodeData { space: AddressSpace::Register, offset: 0x104, size: 4 },
            VarnodeData { space: AddressSpace::Register, offset: 0x100, size: 4 },
        ];
        let mut out = ParamListStandardOut::new();
        out.base.entry_mut().push(make_join_entry(0, 8, 4, &pieces));
        *out.base.entry_mut().last_mut().unwrap().flags_mut() |=
            param_entry_flags::FIRST_STORAGE;
        out.base.entry_mut().push(make_entry(
            1, AddressSpace::Register, 0x200, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY
                | param_entry_flags::FIRST_STORAGE,
        ));
        out.base.set_num_group(2);
        let mut active = ParamActive::new(false);
        active.register_trial_in_space(AddressSpace::Register, Address::new(0x100), 4);
        active.register_trial_in_space(AddressSpace::Register, Address::new(0x104), 4);
        for i in 0..active.get_num_trials() {
            active.get_trial_mut(i).mark_active();
        }
        run_fb(&out, &mut active, false);
    }

    // ---- case 5: fallback, cross-space join pieces ---------------------
    println!("case=fb_cross_space_join");
    {
        let pieces = [
            VarnodeData { space: AddressSpace::Ram, offset: 0x204, size: 4 },
            VarnodeData { space: AddressSpace::Register, offset: 0x100, size: 4 },
        ];
        let mut out = ParamListStandardOut::new();
        out.base.entry_mut().push(make_join_entry(0, 8, 4, &pieces));
        *out.base.entry_mut().last_mut().unwrap().flags_mut() |=
            param_entry_flags::FIRST_STORAGE;
        out.base.set_num_group(1);
        let mut active = ParamActive::new(false);
        active.register_trial_in_space(AddressSpace::Register, Address::new(0x100), 4);
        active.register_trial_in_space(AddressSpace::Ram, Address::new(0x204), 4);
        active.register_trial_in_space(AddressSpace::Stack, Address::new(0x204), 4);
        for i in 0..active.get_num_trials() {
            active.get_trial_mut(i).mark_active();
        }
        run_fb(&out, &mut active, false);
    }

    // ---- case 6: firstOnly skips non-first single-group exclusions -----
    println!("case=fb_first_only_skip");
    {
        let mut out = ParamListStandardOut::new();
        out.base.entry_mut().push(make_entry(
            0, AddressSpace::Register, 0x100, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY
                | param_entry_flags::FIRST_STORAGE,
        ));
        // e1: NOT first_storage, exclusion (alignment 0), one group
        out.base.entry_mut().push(make_entry(
            1, AddressSpace::Register, 0x108, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        out.base.set_num_group(2);
        let mut active = ParamActive::new(false);
        active.register_trial_in_space(AddressSpace::Register, Address::new(0x108), 4);
        active.get_trial_mut(0).mark_active();
        run_fb(&out, &mut active, true);
    }

    // ---- case 7: same layout, firstOnly=false reaches e1 ----------------
    println!("case=fb_first_only_allowed");
    {
        let mut out = ParamListStandardOut::new();
        out.base.entry_mut().push(make_entry(
            0, AddressSpace::Register, 0x100, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY
                | param_entry_flags::FIRST_STORAGE,
        ));
        out.base.entry_mut().push(make_entry(
            1, AddressSpace::Register, 0x108, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        out.base.set_num_group(2);
        let mut active = ParamActive::new(false);
        active.register_trial_in_space(AddressSpace::Register, Address::new(0x108), 4);
        active.get_trial_mut(0).mark_active();
        run_fb(&out, &mut active, false);
    }
}
